src/wrapper/strings/java_str.rs

```diff
@@ -27,11 +90,56 @@ impl<'a: 'b, 'b> JavaStr<'a, 'b> {
         Ok(java_str)
     }
 
-    /// Extract the raw C string pointer from the JavaStr. This will be
-    /// encoded using the JVM internal `CESU-8`-style.
+    /// Get the raw string pointer from the JavaStr.
+    ///
+    /// The string will be `NULL` terminated and encoded as
+    /// [Modified UTF-8](https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8) /
+    /// [CESU-8](https://en.wikipedia.org/wiki/CESU-8).
     pub fn get_raw(&self) -> *const c_char {
         self.internal
     }
+
+    /// Consumes the `JavaStr`, returning the raw string pointer
+    ///
+    /// The string will be `NULL` terminated and encoded as
+    /// [Modified UTF-8](https://en.wikipedia.org/wiki/UTF-8#Modified_UTF-8) /
+    /// [CESU-8](https://en.wikipedia.org/wiki/CESU-8).
+    ///
+    /// # Warning
+    /// The programmer is responsible for making sure the backing string gets
+    /// released when they are done with it, for example by reconstructing a
+    /// [JavaStr] with [`Self::from_raw`], which will release the backing string
+    /// when it is dropped.
+    pub fn into_raw(self) -> *const c_char {
+        let _dont_call_drop = std::mem::ManuallyDrop::new(self);
```

Comment:
Would this not cause a memory leak?

At least, I think so. Since the `JavaStr` is never dropped anywhere now, I think?

Might make more sense to let the user handle the ManuallyDrop?

Reply:
I dont think this introduces a memory leak, rather it avoids a double free...

The JavaStr itself is still dropped it's just that Drop won't be run. Since only thing the Drop implementation frees is the string that we are trying to return to the caller, that's what we want here.

Previously we were returning a pointer to a string that would then get released by the Drop implementation before `into_raw` returned (I.e a pointer to memory that has been freed by the JNI implementation) and that would then lead to a double free.

I think the only thing that can be leaked here is the string we return in case they dont later call `from_raw` to ensure it gets properly dropped but that's made clear in the API docs.